modules/apps/depot/depot-web/src/main/java/com/liferay/depot/web/internal/portlet/action/UpdateMembershipsMVCActionCommand.java

```diff
@@ -113,6 +134,55 @@ private long[] _getGroupIds(PortletRequest portletRequest, User user) {
 		return ArrayUtil.toLongArray(groupIds);
 	}
 
+	private void _validateGroups(ActionRequest actionRequest)
+		throws PortalException {
+
+		Set<Long> groupIds = new HashSet<>();
+
+		for (long addDepotGroupId :
+				StringUtil.split(
+					ParamUtil.getString(actionRequest, "addDepotGroupIds"),
+					0L)) {
+
+			groupIds.add(addDepotGroupId);
+		}
+
+		for (long deleteDepotGroupId :
+				StringUtil.split(
+					ParamUtil.getString(actionRequest, "deleteDepotGroupIds"),
+					0L)) {
+
+			groupIds.add(deleteDepotGroupId);
+		}
+
+		ThemeDisplay themeDisplay = (ThemeDisplay)actionRequest.getAttribute(
+			WebKeys.THEME_DISPLAY);
+
+		PermissionChecker permissionChecker =
+			themeDisplay.getPermissionChecker();
+
+		for (long groupId : groupIds) {
+			Group group = _groupLocalService.fetchGroup(groupId);
+
+			if ((group != null) && group.isDepot() &&
+				!permissionChecker.isCompanyAdmin() &&
+				!GroupPermissionUtil.contains(
+					permissionChecker, group, ActionKeys.ASSIGN_MEMBERS)) {
+
+				throw new PrincipalException.MustHavePermission(
+					themeDisplay.getPermissionChecker(), Group.class.getName(),
+					group.getGroupId(), ActionKeys.ASSIGN_MEMBERS,
+					ActionKeys.ASSIGN_MEMBERS);
+			}
+		}
```

Comment:
Isn't this duplicating `_getGroupIds`? Why not make `_validateGroupIds` take a `long[]` and invoke it like this in the `doProcessAction` method:
```
long[] groupIds = _getGroupIds(...);

_validateGroupIds(groupIds);

....
```

Reply:
no, this is not the same.
in one you are adding all the groups that will be modified.

and on the other you add and delete to the group list.

Reply:
Oh, I see, I hadn't noticed the `user.getGroupIds()`.